crates/core/src/network/transport_connector.rs

```diff
@@ -67,15 +67,13 @@ impl TransportConnect for GrpcConnector {
 
         trace!("Attempting to connect to node {} at {}", node_id, address);
         // Do we have a channel in cache for this address?
-        let channel = {
-            if let dashmap::Entry::Vacant(entry) = self.channel_cache.entry(address.clone()) {
-                let channel = create_tonic_channel_from_advertised_address(address)
-                    .map_err(|e| NetworkError::BadNodeAddress(node_id.into(), e))?;
-                entry.insert(channel.clone());
-                channel
-            } else {
-                self.channel_cache.get(&address).unwrap().clone()
-            }
+        let channel = match self.channel_cache.get(&address) {
+            Some(channel) => channel.clone(),
+            None => self
+                .channel_cache
+                .entry(address.clone())
+                .or_insert_with(|| create_tonic_channel_from_advertised_address(address))
+                .clone(),
```

Comment:
Wouldn't

```
let channel = self
            .channel_cache
            .entry(address.clone())
            .or_insert_with(|| create_tonic_channel_from_advertised_address(address))
            .clone();
```

be sufficient here (without the `get` call before)?

Reply:
See description:
> I've also changed it slightly such that in the case where the channel already exists, we only take a read reference, and only fall back to a write lock when its not present.

This is just a fast-path optimisation, not required to fix the deadlock

Reply:
Ah ok, makes sense. Thanks for the clarification :-)